_algorithm_design/on_computing_next_greater_intervals_in_linear_time.md

```diff
@@ -0,0 +1,179 @@
+
+### Ok, but what about the rightBound?
+
+Consider the `pop` calls in the specified algorithm along with the element that is popped out. Clearly its left bound is the result of the `top` call but what is its right bound? Are you ready for the revelation? Hold tight onto your chairs and breathe deeply because this is going to blow your mind:
+.
+.
+.
+.
+`item`.
+
+By the condition on `item` in the algorithm, the execution will reach the `pop` call if and only if the invariant `list[topo.get_next()] < item` is satisfied. (assume that the stack is not empty in this case, otherwise the right bound index is N)
+
+And we're done. The left bound of the popped element is the result of the `top` call while the right bound is `item`. Since we know the left bound and right bound of an item when it is popped, it is best if we perform any `visit(element, leftBound, rightBound)` operations at this point. [NOTE: With appropriate design choice, one can write the visit routine with the given signature. : END OF NOTE]
+
+### Final algorithm
+
+{% highlight python %}
+topo = MonotonicStack()
+for i, item in enumerate(list):
+    if topo.is_empty() or list[topo.get_next()] > item:
+        topo.push(i)
+    else:
+        while not topo.is_empty() && list[topo.get_next()] < item:
+            element = topo.get_next()
+            topo.pop()
+            leftBound = topo.get_next() if not topo.is_empty() else -1
+            rightBound = i
+            visit(element, leftBound, rightBound)
+        topo.push(i)
+
+while not topo.is_empty():
+    element = topo.get_next()
+    topo.pop()
+    leftBound = topo.get_next() if not topo.is_empty() else -1
+    rightBound = len(list)
+    visit(element, leftBound, rightBound)
+{% endhighlight %}
+
+## Generalization
+
+We saw in the discussion so far how the $>$ operator dicatates the ancestry topology for every element. The idea of the topology easily extends to $<$ operator. What's more? It works out just as elegantly for any binary associative comparison operator used to compare records. As long as your operator can decide an ordering between any two records in the list unambiguosly, we will be able the compute the $nextGreaterInterval$ for every element in linear time.
+
+## Example problems
+
+The well-known [problem][spoj-histogra] that uses the monotonic stack is that of finding the largest rectangle in a histogram. The key idea here is to maintain the $nextGreaterInterval$ with respect to the $<$ operator. `visit` operation computes the area of the rectangle using the height of the popped out element and its $nextGreaterInterval$ and updates the max value of rectangle.
+
+Here's another [problem][cf-max-xor-secondary] that can be solved using the idea of monotonic stack that we just developed. I'll leave this as an exercise for the reader to figure out the specifics.
+
+## Concluding Remarks
```

Comment:
Why the fuck there is no mention of that ML algorithm you said for the errors or something

Reply:
Turns out there's an even simpler way to solve it and I was overstating the use case. So....